Add Kafka Component

g7ed6e opened the pull request:
Relates to #884

g7ed6e committed:
Add Aspire.Confluent.Kafka component

apply pr suggestions

apply pr suggestions

apply pr suggestions

Sort ConfigurationSchema.json properties

Update ConfigurationSchema.json using ConfigSchemaGenerator

apply pr suggestions

apply pr suggestions

apply pr suggesstions

apply pr suggesstions

apply pr suggestions

apply pr suggestions

drop kafka sample from this repo

apply pr suggestions

gfoidl commented on samples/kafkaBasic/Producer/ProducerWorker.cs:
```diff
@@ -0,0 +1,27 @@
+{
+    protected override Task ExecuteAsync(CancellationToken stoppingToken)
+    {
+        return Task.Factory.StartNew(async() =>
```
With the `PeriodicTimer` this is not needed, as for the first time `await timer.WaitForNextTickAsync` is awaited it yields from the current task.
So in essence you create a new thread (due to LongRunning) then exit that thread at the await.

For the `BackgroundService` you can just use `await Task.Yield();` at method entry and avoid the `Task.Factory.StartNew`.

g7ed6e replied:
Done. Thanks!

gfoidl commented on src/Aspire.Hosting/Kafka/KafkaBuilderExtensions.cs:
```diff
@@ -0,0 +1,50 @@
+    /// <returns>A reference to the <see cref="IResourceBuilder{KafkaContainerResource}"/></returns>
+    public static IResourceBuilder<KafkaContainerResource> AddKafkaBrokerContainer(this IDistributedApplicationBuilder builder, string name)
+    {
+        const int kafkaBrokerPort = 9092;
```
Should this be optionally configurable?
Does anyone change the default port anyway?

g7ed6e replied:
Host port is now configurable.

davidfowl commented on samples/kafkaBasic/Producer/ProducerWorker.cs:
```diff
@@ -0,0 +1,25 @@
+
+namespace Producer;
+
+internal sealed class ProducerWorker(IServiceProvider serviceProvider, ILogger<ProducerWorker> logger) : BackgroundService
```
```suggestion
internal sealed class ProducerWorker(IProducer<string, string> producer, ILogger<ProducerWorker> logger) : BackgroundService
```

davidfowl commented on samples/kafkaBasic/Producer/ProducerWorker.cs:
```diff
@@ -0,0 +1,25 @@
+    {
+        await Task.Yield();
+        
+        IProducer<string, string> producer = serviceProvider.GetRequiredService<IProducer<string, string>>();
```
This could be injected right?

```suggestion
```

g7ed6e replied:
yes indeed. good catch

davidfowl commented on src/Aspire.Hosting/Kafka/KafkaBuilderExtensions.cs:
```diff
@@ -0,0 +1,55 @@
+                 Image = "confluentinc/confluent-local",
+                 Tag = "latest"
+             })
+             .WithEnvironment("KAFKA_LISTENERS", "PLAINTEXT://localhost:29092,CONTROLLER://localhost:29093,PLAINTEXT_HOST://0.0.0.0:9092")
+             .WithEnvironment("KAFKA_ADVERTISED_LISTENERS", $"PLAINTEXT://localhost:29092,PLAINTEXT_HOST://localhost:{port}");
```
Where do these ports come from? Is it worth leaving a comment about them?

g7ed6e replied:
I added comments and links. These ports are the defaults of the `confluentinc/confluent-local` image which are set when the container starts if the env var is not supplied. See https://github.com/confluentinc/kafka-images/blob/master/local/include/etc/confluent/docker/configureDefaults

davidfowl commented on samples/kafkaBasic/Consumer/ConsumerWorker.cs:
```diff
@@ -0,0 +1,21 @@
+{
+    protected override Task ExecuteAsync(CancellationToken stoppingToken)
+    {
+        return Task.Factory.StartNew(() =>
+        {
+            consumer.Subscribe("topic");
+            while (!stoppingToken.IsCancellationRequested)
+            {
+                var result = consumer.Consume(stoppingToken);
+                logger.LogInformation($"Received message '{result.Message.Value}'.");
+            }
+        }, stoppingToken, TaskCreationOptions.LongRunning, TaskScheduler.Current);
```
This API is very strange.. no async? Is it blocking?

g7ed6e replied:
Indeed it is. See https://github.com/confluentinc/confluent-kafka-dotnet/blob/master/examples/Consumer/Program.cs and https://docs.confluent.io/kafka-clients/dotnet/current/overview.html#the-consume-loop

gfoidl replied:
Confluentic has bad [API](https://docs.confluent.io/platform/current/clients/confluent-kafka-dotnet/_site/api/Confluent.Kafka.IConsumer-2.html#Confluent_Kafka_IConsumer_2_Consume_CancellationToken_) here. It prefetches messages (if available), but it will block if nothing is there. That would be a perfect fit for `ValutTask` 😉.

Cf. https://github.com/confluentinc/confluent-kafka-dotnet/issues/487

davidfowl commented on samples/kafkaBasic/Producer/ProducerWorker.cs:
```diff
@@ -0,0 +1,24 @@
+{
+    protected override async Task ExecuteAsync(CancellationToken stoppingToken)
+    {
+        await Task.Yield();
```
Not needed here.
```suggestion
```

g7ed6e replied:
done

davidfowl commented on samples/kafkaBasic/Producer/ProducerWorker.cs:
```diff
@@ -0,0 +1,24 @@
+    {
+        await Task.Yield();
+        
+        using PeriodicTimer timer = new PeriodicTimer(TimeSpan.FromSeconds(1));
```
nit:
```suggestion
        using var timer = new PeriodicTimer(TimeSpan.FromSeconds(1));
```

davidfowl commented on src/Components/Aspire.Kafka.Consumer/AspireKafkaConsumerExtensions.cs:
```diff
@@ -0,0 +1,87 @@
+            return consumerBuilder;
+        }
+
+        if (serviceKey is null)
```
Are there any other interesting types to add to the container?

g7ed6e replied:
I added `ProducerConfig` and `ConsumerConfig` to ease unit testing. Also added `Action<ProducerBuilder<TKey, TValue>>` and `Action<ConsumerBuilder<TKey, TValue>>` to only rely on DI to new the `XXXBuilder<TKey, TValue>` types.

davidfowl commented on src/Components/Aspire.Kafka.Consumer/AspireKafkaConsumerExtensions.cs:
```diff
@@ -0,0 +1,87 @@
+        AddKafkaConsumer(builder, $"{DefaultConfigSectionName}:{name}", configureConsumerBuilder, configureConsumerConfig, connectionName: name, serviceKey: name);
+    }
+
+    private static void AddKafkaConsumer<TKey, TValue>(
```
Any health checks, telemetry etc?

g7ed6e replied:
Not yet. I'm thinking to pull logs and metrics from librdkafka using the below delegates that can be registered within `ProducerBuilder<TKey, TValue>` and `ConsumerBuilder<TKey, TValue>` [SetLogHandler](https://docs.confluent.io/platform/current/clients/confluent-kafka-dotnet/_site/api/Confluent.Kafka.ConsumerBuilder-2.html#Confluent_Kafka_ConsumerBuilder_2_SetLogHandler_System_Action_Confluent_Kafka_IConsumer__0__1__Confluent_Kafka_LogMessage__), [SetErrorHandler](https://docs.confluent.io/platform/current/clients/confluent-kafka-dotnet/_site/api/Confluent.Kafka.ConsumerBuilder-2.html#Confluent_Kafka_ConsumerBuilder_2_SetErrorHandler_System_Action_Confluent_Kafka_IConsumer__0__1__Confluent_Kafka_Error__), [SetStatisticsHandler](https://docs.confluent.io/platform/current/clients/confluent-kafka-dotnet/_site/api/Confluent.Kafka.ConsumerBuilder-2.html#Confluent_Kafka_ConsumerBuilder_2_SetStatisticsHandler_System_Action_Confluent_Kafka_IConsumer__0__1__System_String__). The [STATISTICS.json](https://github.com/confluentinc/librdkafka/blob/master/STATISTICS.md) contains metrics.

gfoidl commented on samples/kafkaBasic/Producer/ProducerWorker.cs:
```diff
@@ -0,0 +1,22 @@
+    protected override async Task ExecuteAsync(CancellationToken stoppingToken)
+    {
+        using var timer = new PeriodicTimer(TimeSpan.FromSeconds(1));
+        int i = 0;
```
```suggestion
        long i = 0;
```
It's just a demo, but should it be `long` to prevent potential overflow?
And as demo it should showcase good conduct.

g7ed6e replied:
done

g7ed6e commented on src/Aspire.Hosting/Kafka/KafkaBuilderExtensions.cs:
```diff
@@ -0,0 +1,58 @@
+    /// <returns>A reference to the <see cref="IResourceBuilder{KafkaContainerResource}"/></returns>
+    public static IResourceBuilder<KafkaContainerResource> AddKafkaBrokerContainer(this IDistributedApplicationBuilder builder, string name, int? port = null)
+    {
+        const int kafkaBrokerPort = 9092;
```
done

mitchdenny commented:
Overall, the app model side of this is looking pretty good. I think we need some end to end integration tests for this.

davidfowl commented on src/Components/Aspire.Kafka.Consumer/IConsumerConnection.cs:
```diff
@@ -0,0 +1,19 @@
+/// </summary>
+/// <typeparam name="TKey">The type of the key of a message.</typeparam>
+/// <typeparam name="TValue">The type of the value of a message.</typeparam>
+public interface IConsumerConnection<TKey, TValue>
```
We're trying really hard to avoid abstractions owned by the aspire components.

g7ed6e replied:
I see. I'm trying to avoid collisions with types not owned by aspire that may have been registered in DI by user/customer code. I could have made these new types internal but then unit testing would became painful as once built `IProducer<TKey, TValue>` and `IConsumer<TKey, TValue>` does not expose their configuration values.

mitchdenny commented:
Once health checks are implemented, please wire-up the Kafa component inside our smoke test (new):

1. Add container to app host here: https://github.com/dotnet/aspire/blob/60bd2b0b5c29ea02507265199070d4dac1fad712/tests/testproject/TestProject.AppHost/TestProgram.cs#L18
2. Add component to the integration test project: https://github.com/dotnet/aspire/blob/60bd2b0b5c29ea02507265199070d4dac1fad712/tests/testproject/TestProject.IntegrationServiceA/Program.cs#L5

g7ed6e commented:
@mitchdenny Can you add `Confluent.Kafka` 2.3.0 and `AspNetCore.HealthChecks.Kafka` 7.0.0 ?

davidfowl commented:
I think we want a single component for kafka, not a package for the consumer and produce. Also, the component should be Aspire.Confluent.Kafka since that's the name of the underlying client library we're using. 

PS: People are out of vacation this week so it's slow 😄

riverar commented:
Side comment: As additional components get developed/added, I wonder if we'll need an _ingredients label_ so developers know which components (and which versions) do and do not have health checks, telemetry, etc.

mitchdenny commented:
> I wonder if we'll need an _ingredients label_ so developers know which components (and which versions) do and do not have health checks, telemetry, etc.

We probably shouldn't be merging it if it doesn't have telemetry, health-checks etc. In terms of dependency graph, I think we get that for free from NuGet so I don't think we need to do anything extra there.

mitchdenny commented:
> @mitchdenny Can you add `Confluent.Kafka` 2.3.0 and `AspNetCore.HealthChecks.Kafka` 7.0.0 ?

They should be in dotnet-public shortly.

eerhardt reviewed (commented):
We should also add a README and ConfigurationSchema.json files

riverar commented on src/Aspire.Hosting/Kafka/KafkaBuilderExtensions.cs:
```diff
@@ -0,0 +1,58 @@
+             .WithAnnotation(new ServiceBindingAnnotation(ProtocolType.Tcp, port: port, containerPort: KafkaBrokerPort))
+             .WithAnnotation(new ContainerImageAnnotation
+             {
+                 Image = "confluentinc/confluent-local",
```
How does the .NET team evaluate container image defaults like this? How do we know it's safe to use? Is Microsoft signing off on this as trusted? Are there license, support, privacy, import/export implications?

mitchdenny replied:
The .NET team cannot evaluate the underlying third-party containers for which we have API support for suitability for specific customer requirements.

You should evaluate whichever containers you chose to compose and automate with Aspire to ensure they meet your, your employers or your government’s requirements on security and safety as well as cryptographic regulations and any other regulatory or corporate stands that may apply to your use of the container.

riverar replied:
@mitchdenny Mitch, my concerns aren’t about the _specific customer requirements_. Rather, I’m asking about some fundamental safeguards. It’s quite unsettling that my concerns seem to have been deliberately dismissed with a legal response.

https://github.com/dotnet/aspire/blob/main/README.md#note-on-containers-used-by-aspire-componentsresources

mitchdenny replied:
Appreciate the feedback and sorry for the delay in responding. I've been OOF and haven't managed to get around to everything. I'm sorry you feel that we are dismissing your concerns.

When we say the .NET team cannot evaluate the underlying third-party containers, we mean that ultimately because we are not in control of those third-party container build pipelines, we can't attest to what is in them. Its conceptually the same as referencing one of those containers in a Docker Compose file, Kubernetes deployment or similar container orchestrator. We don't distribute the containers (with a few exceptions).

That said, I do think you've raised some good points. For example, perhaps we should consider referencing the containers we depend on via their SHA256 identifier rather than a tag. But there is a negative flipside to that in that lets say we reference a container via its SHA, and that particular version has a vulnerability then not using a floating tag (either latest or something a bit more version/channel orientated) willl mean you don't automatically get patched.

This is a pretty naunced topic so I would encourage you to open up a specific issue for us to discuss rather than side track this PR.

g7ed6e commented on src/Components/Aspire.Confluent.Kafka/ConsumerConnectionFactory.cs:
```diff
@@ -0,0 +1,32 @@
+/// </summary>
+/// <typeparam name="TKey">The type of the key of a message.</typeparam>
+/// <typeparam name="TValue">The type of the value of a message.</typeparam>
+public class ConsumerConnectionFactory<TKey, TValue>
```
Once built, neither `IConsumer<TKey, TValue>` nor `ConsumerBuilder<TKey, TValue>` expose the values used to build. We need this type public to fetch the config used in unit tests. (same thing for ConsumerConnectionFactory).

g7ed6e replied:
@eerhardt would you accept using reflection + compiled expression to access configured values on unit tests ? If the answer is yes we should be able to move the visibility of that type to internal. 
Also I have the same need to unit tests metrics produced by this component as metrics  are produced by internal types.
/cc @davidfowl

g7ed6e commented on src/Components/Aspire.Confluent.Kafka/ConfluentKafkaMetrics.cs:
```diff
@@ -0,0 +1,90 @@
+
+    public static class Counters
+    {
+        public const string ReplyQueue = "replyq";
```
I'm not satisfied either by these names (basically same names as the ones from librdkafka [statistics.json](https://github.com/confluentinc/librdkafka/blob/master/STATISTICS.md)), but the naming pattern you linked seems to concern spans not metrics.
As these constants are used to expose specific metrics related to a given producer or consumer instance, I think we should instead follow a metric naming pattern.

davidfowl replied:
@lmolkova you are needed!

g7ed6e replied:
for the time being I prefixed metrics with `messaging.kafka`

lmolkova replied:
Kafka metrics are in a weird place.
OTel is in the process of adding general messaging metrics into the spec - https://github.com/open-telemetry/semantic-conventions/pull/163 (will be merged soon).

However Kafka as a project decided (AFAIK) to report a set of client metrics ([KIP-714](https://cwiki.apache.org/confluence/display/KAFKA/KIP-714%3A+Client+metrics+and+observability)) that would be sent by default back to the broker and users would consume them from it. They decided not to follow otel semconv for the time being because of some back-compat reasons and limitations of existing metric impl on the broker side. Since they don't follow otel conventions, they also don't prefix with `messaging.` namespace.

I have a couple of suggestion 
- let's get an opinion from Kafka folks to understand the bigger picture.  @AndrewJSchofield what are your thoughts on Kafka integration component for .NET Aspire emitting it's own set of metrics? Also, could you confirm if I remember the decision (above) correctly or maybe something has changed? I envision the same conversation in OTel instrumentation world in Java and Python.

- in the short term, WRT metrics added here, I'd recommend aligning them with messaging semantic conventions when applicable and following otel naming conventions otherwise. I'll go through the changes and add my suggestions.

/cc @pyohannes

davidfowl replied:
I think we should align with the semantic conventions as well, even if they are not final. Maybe we can help push towards that goal.

lmolkova replied:
> I think we should align with the semantic conventions as well, even if they are not final. Maybe we can help push towards that goal.

we can't because we just don't get the right data from the kafka client library. 

We're going to discuss naming/approach with otel semconv community - https://github.com/open-telemetry/semantic-conventions/issues/578, I should have an update soon.

lmolkova replied:
@g7ed6e would you be able to take a quick look at https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/bbfe950ad0ace8123a5e6817fb3767e27a1a2cee/instrumentation/kafka/kafka-clients/kafka-clients-2.6/library/README.md and check if any of those metrics apply here and if we can unify the naming? (just discovered they exist).

In any case, metric naming and that discussion should not block this PR and can be changed in the follow-ups.

g7ed6e replied:
@pyohannes : I updated and renamed `messaging.kafka.message.transmitted` to [messaging.publish.messages](https://github.com/open-telemetry/semantic-conventions/blob/main/docs/messaging/messaging-metrics.md#metric-messagingpublishmessages) and 
`messaging.kafka.message.received` to [messaging.receive.messages](https://github.com/open-telemetry/semantic-conventions/blob/main/docs/messaging/messaging-metrics.md#metric-messagingreceivemessages).

@lmolkova : I had a look to the linked implementation. Currently (in this Aspire PR) we do not intercept messages between user code and kafka client library. Instead we use a librdkafka feature which perdiodically fetch statistics (please notice we only parse the [top level properties from the statistics json](https://github.com/confluentinc/librdkafka/blob/master/STATISTICS.md) for now, there should be other metrics we may probably use and map to metrics in the future).
Regarding the interceptor / decorator approach we may implement it, but before I would like to ensure that we want to do this in Aspire component as having a look to the other components they do not hold too much code/logic and only exposes existing clients. 

@eerhardt / @davidfowl  : What are your thoughts about implementing otel distributed tracing with an internal wrapper around `IProducer<TKey, TValue>` / `IConsumer<TKey, TValue>` living in Aspire.Confluent.Kafka.

davidfowl replied:
> @eerhardt / @davidfowl : What are your thoughts about implementing otel distributed tracing with an internal wrapper around IProducer<TKey, TValue> / IConsumer<TKey, TValue> living in Aspire.Confluent.Kafka.

Yes, this sounds reasonable as a way to drive the conversation forward for the default .NET Kafka library.

@eerhardt I know we don't do this today with rabbit MQ (I think it's harder there because there are more APIs to wrap).

g7ed6e commented on src/Components/Aspire.Confluent.Kafka/AspireKafkaConsumerExtensions.cs:
```diff
@@ -0,0 +1,177 @@
+
+        if (serviceKey is null)
+        {
+            builder.Services.AddSingleton<ConsumerSettingsRetriever<TKey, TValue>>(() => settings);
```
done

g7ed6e commented:
> > @mitchdenny Can you add `Confluent.Kafka` 2.3.0 and `AspNetCore.HealthChecks.Kafka` 7.0.0 ?
> 
> They should be in dotnet-public shortly.

Thank you. `Confluent.Kafka` is there but `AspNetCore.HealthChecks.Kafka` is not found.

g7ed6e commented on src/Components/Aspire.Confluent.Kafka/KafkaProducerSettings.cs:
```diff
@@ -0,0 +1,43 @@
+
+    public bool HealthChecks { get; set; } = true;
+
+    public ProducerConfig? HealthCheckConfig { get; set; }
```
done

g7ed6e commented:
@eerhardt @davidfowl `Activity` is not implemented in Confluent.Kafka. I'm thinking to wrap `IProducer` and `IConsumer` in a decorator to intercept call and start appropriate `Activity`. However i'm concerned about maintainability. Is it worth it ?

forlack commented:
I am the PM owner over our Kafka Clients at Confluent.  Let me know if you need any help or collaboration from some of our engineers.  Happy to answer any questions we can.  Thanks!

g7ed6e commented:
> I am the PM owner over our Kafka Clients at Confluent. Let me know if you need any help or collaboration from some of our engineers. Happy to answer any questions we can. Thanks!

Thank you @forlack for reaching that PR. The missing piece would be to get [OTel Distributed tracing](https://github.com/open-telemetry/semantic-conventions/blob/main/docs/messaging/README.md) support in Confluent.Kafka. My thinking is that it could be implemented from both the inside and the outside of Confluent.Kafka (however inside would be easier) but having builtin support for this in Confluent.Kafka would benefit to more people and libraries.
I'm linking here the [related Confluent.Kafka issue](https://github.com/confluentinc/confluent-kafka-dotnet/issues/1269).

lmolkova commented on src/Components/Aspire.Confluent.Kafka/ConfluentKafkaMetrics.cs:
```diff
@@ -0,0 +1,91 @@
+    {
+        public const string ClientId = "client_id";
+        public const string Type = "type";
+        public const string Name = "name";
```
if I understand correctly, name is a name of a broker host in the cluster - OTel uses [`server.address` and `server.port`](https://github.com/open-telemetry/semantic-conventions/blob/main/docs/attributes-registry/server.md#server-attributes) for it.

g7ed6e replied:
`name` is the name of the handle of the current producer or consumer in librdkafka. It is a string which concatenate the client_id (which can be set during building the producer or the consumer), a `#` and the identifier of the producer/consumer. 

From the [docs](https://github.com/confluentinc/librdkafka/blob/master/STATISTICS.md)

Field | Type | Example | Description
----- | ---- | ------- | -----------
name | string | `"rdkafka#producer-1"` | Handle instance name

lmolkova commented on src/Components/Aspire.Confluent.Kafka/ConfluentKafkaMetrics.cs:
```diff
@@ -0,0 +1,91 @@
+
+    public static class Tags
+    {
+        public const string ClientId = "client_id";
```
otel attribute is [`messaging.client_id`](https://github.com/open-telemetry/semantic-conventions/blob/main/docs/attributes-registry/messaging.md). 

Also, it's not common to report client Id in the metrics because of potentially high cardinality (if user app creates a new instance every time their app is not only slow, but also now kafka metrics are capped and not useful). 
Instead of adding it as an attribute, could you just add a log whenever new client is created or have a counter for number of active client instances?

g7ed6e replied:
@lmolkova  Updated the PR and renamed client_id attribute to messaging.client_id.

lmolkova commented on src/Components/Aspire.Confluent.Kafka/ConfluentKafkaMetrics.cs:
```diff
@@ -0,0 +1,91 @@
+
+    public static class Tags
+    {
+        public const string ClientId = "client_id";
```
I wonder if the topic name is available - it'd be great to add it and it should go to `messaging.destination.name` attribute

lmolkova commented on src/Components/Aspire.Confluent.Kafka/ConfluentKafkaMetrics.cs:
```diff
@@ -0,0 +1,91 @@
+    public static class Counters
+    {
+        private const string Prefix = "messaging.kafka.";
+        public const string ReplyQueue = Prefix + "replyq";
```
based on the description "Number of ops (callbacks, events, etc) waiting in queue for application to serve with rd_kafka_poll()", this should be called something like 
`messaging.kafka.consumer.queue.message_count` to follow otel semconv


this would be more descriptive and would also remove the need for `type` attribute

g7ed6e replied:
renamed and removed the `type` attribute.

mitchdenny commented:
> @mitchdenny Thank you. `Confluent.Kafka` is there but `AspNetCore.HealthChecks.Kafka` is not found.

Sorry about that, I made a mistake with that import. It is there now.

lmolkova commented on src/Components/Aspire.Confluent.Kafka/ConfluentKafkaMetrics.cs:
```diff
@@ -0,0 +1,91 @@
+    {
+        private const string Prefix = "messaging.kafka.";
+        public const string ReplyQueue = Prefix + "replyq";
+        public const string MessageCount = Prefix + "msg_cnt";
```
this one ("Current number of messages in producer queues") should rather be `messaging.kafka.producer.queue.message_count`

g7ed6e replied:
renamed and removed the `type` attribute.

lmolkova commented on src/Components/Aspire.Confluent.Kafka/ConfluentKafkaMetrics.cs:
```diff
@@ -0,0 +1,91 @@
+        private const string Prefix = "messaging.kafka.";
+        public const string ReplyQueue = Prefix + "replyq";
+        public const string MessageCount = Prefix + "msg_cnt";
+        public const string MessageSize = Prefix + "msg_size";
```
> Current total size of messages in producer queues

`messaging.kafka.producer.queue.size`

g7ed6e replied:
renamed and removed the `type` attribute.

lmolkova commented on src/Components/Aspire.Confluent.Kafka/ConfluentKafkaMetrics.cs:
```diff
@@ -0,0 +1,91 @@
+        public const string ReplyQueue = Prefix + "replyq";
+        public const string MessageCount = Prefix + "msg_cnt";
+        public const string MessageSize = Prefix + "msg_size";
+        public const string Tx = Prefix + "tx";
```
> Total number of requests sent to Kafka brokers

Ideally, we'd want to report request duration histogram here (and count can be derived from it) with name like `messaging.kafka.request.duration`. If duration is not available and we're stuck with counters, then something like `messaging.kafka.requests` would work.

Another option is to report different metrics for producer and consumer and have `messaging.kafka.producer.requests`  and `messaging.kafka.consumer.requests`  since we're going to have producer/consumer in metric names anyway (above comments)

lmolkova commented on src/Components/Aspire.Confluent.Kafka/ConfluentKafkaMetrics.cs:
```diff
@@ -0,0 +1,91 @@
+        public const string MessageCount = Prefix + "msg_cnt";
+        public const string MessageSize = Prefix + "msg_size";
+        public const string Tx = Prefix + "tx";
+        public const string TxBytes = Prefix + "tx_bytes";
```
> Total number of bytes transmitted to Kafka brokers

if we know sizes of individual requests, we should expose this metrics as a histogram (similar to [`http.client.request.body.size`](https://github.com/open-telemetry/semantic-conventions/blob/main/docs/http/http-metrics.md#metric-httpclientrequestbodysize) and call it something like `messaging.kafka.request.body.size`

If we only know aggregated value then 
`messaging.kafka.network.io` (similarly to [`process.network.io`](https://github.com/open-telemetry/semantic-conventions/blob/main/docs/system/process-metrics.md) could be an option if we add `direction` attribute.

Also `messaging.kafka.network.tx`/`messaging.kafka.network.transmitted` and `messaging.kafka.network.rx`/`messaging.kafka.network.received` maybe even a better option to avoid adding a new attributes

g7ed6e replied:
> Also `messaging.kafka.network.tx`/`messaging.kafka.network.transmitted` and `messaging.kafka.network.rx`/`messaging.kafka.network.received` maybe even a better option to avoid adding a new attributes

I opted for this.

lmolkova commented on src/Components/Aspire.Confluent.Kafka/ConfluentKafkaMetrics.cs:
```diff
@@ -0,0 +1,91 @@
+        public const string Tx = Prefix + "tx";
+        public const string TxBytes = Prefix + "tx_bytes";
+        public const string Rx = Prefix + "rx";
+        public const string RxBytes = Prefix + "rx_bytes";
```
see the comment on `tx_bytes`

lmolkova commented on src/Components/Aspire.Confluent.Kafka/ConfluentKafkaMetrics.cs:
```diff
@@ -0,0 +1,91 @@
+        public const string MessageSize = Prefix + "msg_size";
+        public const string Tx = Prefix + "tx";
+        public const string TxBytes = Prefix + "tx_bytes";
+        public const string Rx = Prefix + "rx";
```
see comment on `tx`

lmolkova commented on src/Components/Aspire.Confluent.Kafka/ConfluentKafkaMetrics.cs:
```diff
@@ -0,0 +1,91 @@
+        public const string TxBytes = Prefix + "tx_bytes";
+        public const string Rx = Prefix + "rx";
+        public const string RxBytes = Prefix + "rx_bytes";
+        public const string TxMessages = Prefix + "txmsgs";
```
> Total number of messages transmitted (produced) to Kafka brokers

this is a generic soon-to-be-merged otel metric `messaging.publish.messages` -  https://github.com/open-telemetry/semantic-conventions/pull/163/

lmolkova commented on src/Components/Aspire.Confluent.Kafka/ConfluentKafkaMetrics.cs:
```diff
@@ -0,0 +1,91 @@
+        public const string TxMessages = Prefix + "txmsgs";
+        public const string TxMessageBytes = Prefix + "txmsg_bytes";
+        public const string RxMessages = Prefix + "rxmsgs";
+        public const string RxMessageBytes = Prefix + "rxmsg_bytes";
```
see comment on `tx_bytes`

lmolkova commented on src/Components/Aspire.Confluent.Kafka/ConfluentKafkaMetrics.cs:
```diff
@@ -0,0 +1,91 @@
+        public const string Rx = Prefix + "rx";
+        public const string RxBytes = Prefix + "rx_bytes";
+        public const string TxMessages = Prefix + "txmsgs";
+        public const string TxMessageBytes = Prefix + "txmsg_bytes";
```
> Total number of message bytes (including framing) received from Kafka brokers

I *think* otel will introduce a generic one and would likely follow something like `messaging.publish.message.size`, `messaging.publish.batch.size` pattern. 

Another option would be to have consistency with `messaging.kafka.network.transmitted` and call it something like `messaging.kafka.message.tx`/`messaging.kafka.message.transmitted``

g7ed6e replied:
> Another option would be to have consistency with `messaging.kafka.network.transmitted` and call it something like `messaging.kafka.message.tx`/`messaging.kafka.message.transmitted``

I opted for this.

lmolkova commented on src/Components/Aspire.Confluent.Kafka/ConfluentKafkaMetrics.cs:
```diff
@@ -0,0 +1,91 @@
+        public const string RxBytes = Prefix + "rx_bytes";
+        public const string TxMessages = Prefix + "txmsgs";
+        public const string TxMessageBytes = Prefix + "txmsg_bytes";
+        public const string RxMessages = Prefix + "rxmsgs";
```
> Total number of messages consumed, not including ignored messages (due to offset, etc), from Kafka brokers.

this is a generic soon-to-be-merged otel metric `messaging.receive.messages` - https://github.com/open-telemetry/semantic-conventions/pull/163

davidfowl commented on src/Components/Aspire.Confluent.Kafka/MetricsChannel.cs:
```diff
@@ -0,0 +1,18 @@
+
+internal sealed class MetricsChannel
+{
+    private readonly Channel<string> _channel = Channel.CreateUnbounded<string>(new UnboundedChannelOptions()
```
Feels like this should be bounded or we risk creating a memory leak.

g7ed6e replied:
bounded to arbitrary 10k.

davidfowl commented on src/Components/Aspire.Confluent.Kafka/MetricsService.cs:
```diff
@@ -0,0 +1,78 @@
+            while (_channel.Reader.TryRead(out var json))
+            {
+                Statistics? statistics;
+                try
+                {
+                     statistics = JsonSerializer.Deserialize(json, typeof(Statistics), StatisticsJsonSerializerContext.Default) as Statistics;
+                }
+                catch
+                {
+                    return;
+                }
+
+                if (statistics == null || statistics.Name == null)
+                {
+                    return;
+                }
```
we should make sure this is as efficient as possible.

davidfowl commented on tests/testproject/TestProject.IntegrationServiceA/Program.cs:
```diff
@@ -2,11 +2,13 @@
 builder.AddRedis("redis");
 builder.AddNpgsqlDataSource("postgres");
 builder.AddRabbitMQ("rabbitmq");
+builder.AddKafkaProducer<string, string>("kafka");
```
Today this test relies on the health check to verify that things are working

g7ed6e replied:
@eerhardt I added an integration test here f0c4d03e04b5640937c0055e930e7190068e9f91

g7ed6e commented on src/Components/Aspire.Confluent.Kafka/MetricsService.cs:
```diff
@@ -0,0 +1,78 @@
+                Statistics? statistics;
+                try
+                {
+                     statistics = JsonSerializer.Deserialize(json, typeof(Statistics), StatisticsJsonSerializerContext.Default) as Statistics;
```
done

lachmatt commented:
hi @g7ed6e, it looks like some parts of it (e.g. metrics) could be extracted into instrumentation library, which could then be referenced from here (similarly to how `StackExchange.Redis` [component](https://github.com/dotnet/aspire/tree/main/src/Components/Aspire.StackExchange.Redis) references instrumentation [library](https://github.com/open-telemetry/opentelemetry-dotnet-contrib/tree/main/src/OpenTelemetry.Instrumentation.StackExchangeRedis)). I know this requires additional effort, but I think it would benefit OTel community.
Have you considered that approach?

g7ed6e commented:
> hi @g7ed6e, would it be possible to extract e.g. metrics into instrumentation library, which could then be referenced from here (similarly to how `StackExchange.Redis` [component](https://github.com/dotnet/aspire/tree/main/src/Components/Aspire.StackExchange.Redis) references instrumentation [library](https://github.com/open-telemetry/opentelemetry-dotnet-contrib/tree/main/src/OpenTelemetry.Instrumentation.StackExchangeRedis))? 
> 
> I know this requires additional effort, but I think it would benefit OTel community.
> 
> Have you considered that approach?

Hi @lachmatt I agree it looks like the best long term approach. I will open an issue and a PR there.

mitchdenny commented on src/Aspire.Dashboard/Resources/Columns.Designer.cs:
```diff
@@ -1,10 +1,10 @@
 ﻿//------------------------------------------------------------------------------
 // <auto-generated>
-//     This code was generated by a tool.
```
These should be in English. Not sure whey they are changed in this PR?

sebastienros replied:
Always because of the French

g7ed6e replied:
done. I did not noticed my VS2022 preview was french ;). It would be great that default VS installer use English instead of local language.

eerhardt commented on src/Components/Aspire.Confluent.Kafka/ConfluentKafkaCommon.cs:
```diff
@@ -0,0 +1,16 @@
+
+internal sealed class ConfluentKafkaCommon
+{
+    public const string MeterName = "Aspire.Confluent.Kafka";
```
Why does this have an `Aspire` prefix? When we take https://github.com/open-telemetry/opentelemetry-dotnet-contrib/pull/1493, it won't have a prefix right? Will that be a breaking change?

eerhardt commented on src/Components/Aspire.Confluent.Kafka/ConfluentKafkaMetrics.cs:
```diff
@@ -0,0 +1,109 @@
+{
+    private readonly IMeterFactory _meterFactory;
+    private readonly Meter _meter;
+    private static readonly Dictionary<string, string> s_descriptions = new Dictionary<string, string>
```
Do we really need this Dictionary? Can't these just be `const string`s? They are only used in 1 place.

g7ed6e replied:
done

eerhardt commented on src/Components/Aspire.Confluent.Kafka/KafkaProducerSettings.cs:
```diff
@@ -0,0 +1,41 @@
+
+namespace Aspire.Confluent.Kafka;
+
+public sealed class KafkaProducerSettings
```
Can you add XML docs to all public APIs?

g7ed6e replied:
done

eerhardt commented on src/Components/Aspire.Confluent.Kafka/HealthCheckExtensions.cs:
```diff
@@ -0,0 +1,32 @@
+
+namespace Aspire.Confluent.Kafka;
+
+internal static class HealthCheckExtensions
```
You shouldn't need to copy this file. You can just source link to the `Common` folder.

g7ed6e replied:
done

eerhardt commented on src/Components/Aspire.Confluent.Kafka/README.md:
````diff
@@ -0,0 +1,149 @@
+The `WithReference` method configures a connection in the `MyService` project named `messaging`. In the _Program.cs_ file of `MyService`, the Apache Kafka broker connection can be consumed using:
+
+```csharp
+builder.AddKafkaConnection("messaging");
````
This doesn't seem to be a valid method. Did you mean `AddKafkaConsumer`?

eerhardt commented on src/Components/Aspire.Confluent.Kafka/AspireKafkaConsumerExtensions.cs:
```diff
@@ -0,0 +1,175 @@
+                    options.Configuration.SocketTimeoutMs = 1000;
+                    options.Configuration.MessageTimeoutMs = 1000;
+                    options.Configuration.StatisticsIntervalMs = 0;
+                    return new KafkaHealthCheck(options);
```
Instead of establishing a new connection, should the health check use the IProducer we registered in DI? See 

https://github.com/dotnet/aspire/blob/23e168d898f8ab8ad3eaafa2db7d6af47652dd78/src/Components/README.md?plain=1#L96

g7ed6e replied:
The `AspNetCore.HealthChecks.Kafka` package builds its own producer instance. 
If it was implemented, another limitation would be at consumer side we would not be able to reuse the client instance as these are different instances.

eerhardt commented on src/Components/Aspire.Confluent.Kafka/AspireKafkaConsumerExtensions.cs:
```diff
@@ -0,0 +1,175 @@
+        {
+            string healthCheckName = serviceKey is null
+                ? ConfluentKafkaCommon.ConsumerHealthCheckName
+                : string.Format(CultureInfo.InvariantCulture, ConfluentKafkaCommon.KeyedConsumerHealthCheckNameFormat, connectionName);
```
Could just do a simple concat here instead of string.Format.

g7ed6e replied:
done

eerhardt commented on src/Components/Aspire.Confluent.Kafka/KafkaProducerSettings.cs:
```diff
@@ -0,0 +1,41 @@
+
+        if (ConnectionString is not null)
+        {
+            Config!.BootstrapServers = ConnectionString;
```
```suggestion
            Config.BootstrapServers = ConnectionString;
```

You've already asserted it is not null. (applies in other places as well)

g7ed6e replied:
done

eerhardt commented on tests/Aspire.Confluent.Kafka.Tests/ReflectionHelpers.cs:
```diff
@@ -0,0 +1,16 @@
+
+internal static class ReflectionHelpers
+{
+    public static readonly Lazy<Assembly> ComponentAssembly = new Lazy<Assembly>(() => Assembly.Load("Aspire.Confluent.Kafka"));
```
```suggestion
    public static readonly Assembly ComponentAssembly = typeof(AspireKafkaConsumerExtensions).Assembly;
```

g7ed6e replied:
done

eerhardt commented on tests/Aspire.Confluent.Kafka.Tests/ReflectionHelpers.cs:
```diff
@@ -0,0 +1,16 @@
+internal static class ReflectionHelpers
+{
+    public static readonly Lazy<Assembly> ComponentAssembly = new Lazy<Assembly>(() => Assembly.Load("Aspire.Confluent.Kafka"));
+    public static readonly Lazy<Type?> MetricsChannelType = new Lazy<Type?>(() => ComponentAssembly.Value.GetType("Aspire.Confluent.Kafka.MetricsChannel"));
```
If you made this

```suggestion
    public static readonly Lazy<Type> MetricsChannelType = new Lazy<Type>(() => ComponentAssembly.Value.GetType("Aspire.Confluent.Kafka.MetricsChannel")!);
```

Then you wouldn't need `!` operators everywhere else.

g7ed6e replied:
done. thanks!

eerhardt commented on tests/Aspire.Hosting.Tests/DistributedApplicationTests.cs:
```diff
@@ -423,6 +423,38 @@ public async Task VerifyDockerAppWorks()
     }
 
+    [LocalOnlyFact]
+    public async Task KafkaComponentCanProduceAndConsume()
```
Checkout https://github.com/dotnet/aspire/pull/1217. We are refactoring how these tests are written. Once that PR is merged, you can merge with it and adjust the tests as necessary.

g7ed6e replied:
done

eerhardt reviewed (commented):
Thanks, @g7ed6e for this contribution. It is really starting to take shape!

Do you plan on getting https://github.com/open-telemetry/opentelemetry-dotnet-contrib/pull/1493 done before this PR can be merged? Or do you want to put the code in now, and when the OTel change is done, take the code out of here and use the official OTel support?

g7ed6e commented on src/Aspire.Dashboard/Resources/Columns.Designer.cs:
```diff
@@ -1,10 +1,10 @@
 ﻿//------------------------------------------------------------------------------
 // <auto-generated>
-//     This code was generated by a tool.
```
done

g7ed6e commented on tests/testproject/TestProject.IntegrationServiceA/appsettings.json:
```diff
@@ -5,5 +5,23 @@
   },
-  "AllowedHosts": "*"
+  "AllowedHosts": "*",
+  "Aspire": {
```
done. I hardcoded the required configuration in the test project.

g7ed6e commented:
> Thanks, @g7ed6e for this contribution. It is really starting to take shape!
> 
> Do you plan on getting [open-telemetry/opentelemetry-dotnet-contrib#1493](https://github.com/open-telemetry/opentelemetry-dotnet-contrib/pull/1493) done before this PR can be merged? Or do you want to put the code in now, and when the OTel change is done, take the code out of here and use the official OTel support?

Hi @eerhardt would you be open to have this PR merged with healthcheck support only ? (It requires a bit of code cleanup to drop metrics introduced here so please let me know) and then, once otel package is ready, i can open a follow up PR here to integrate traces and metrics support. The main benefit I see is to avoid conflicts and/or breaking changes between the two implementations. The otel implementation will be driven by otel maintainers and I cannot guarantee there won't be breaking changes.

lachmatt commented:
> I am the PM owner over our Kafka Clients at Confluent. Let me know if you need any help or collaboration from some of our engineers. Happy to answer any questions we can. Thanks!

hi @forlack, is there any chance Confluent`s Kafka .NET Client could be instrumented using e.g ActivitySource API to provide tracing support?
It was discussed in the past in https://github.com/confluentinc/confluent-kafka-dotnet/issues/1269

eerhardt commented:
> Hi @eerhardt would you be open to have this PR merged with healthcheck support only ? (It requires a bit of code cleanup to drop metrics introduced here so please let me know) and then, once otel package is ready, i can open a follow up PR here to integrate traces and metrics support. The main benefit I see is to avoid conflicts and/or breaking changes between the two implementations. The otel implementation will be driven by otel maintainers and I cannot guarantee there won't be breaking changes.

I think it would be OK to merge the PR with the current Metrics implementation as it is. It would be a nice way to get people using it and get feedback. .NET Aspire is still in preview, and should be for a few months, so making breaking telemetry changes is acceptable.

Note that we can keep the Kafka Aspire component in "preview" / "unstable" until we get the otel changes in and consumed here.

cc @davidfowl @DamianEdwards - are you OK with that plan?

DamianEdwards commented:
> Note that we can keep the Kafka Aspire component in "preview" / "unstable" until we get the otel changes in and consumed here.

Sounds good.

g7ed6e commented:
Hi @mitchdenny may I ask you to make `AspNetCore.HealthChecks.Kafka` 8.0.0 available to the nuget feeds ?

mitchdenny commented:
@g7ed6e  importing now.

g7ed6e commented:
> @g7ed6e  importing now.

Thanks @mitchdenny CI now completes successfully

eerhardt commented:
I pushed 2 commits to update the ConfigurationSchema.json file to be more complete (sorted and using the generator from #1383). I put them in 2 commits for easy reviewing - first just sorts the existing properties. The second can be diff'd by itself to see what the generator changed.

Unfortunately I hit a bug with the ConfigurationBinder source generator - https://github.com/dotnet/runtime/issues/96652. We aren't able to automate updating the ConfigurationSchema.json file automatically until we have a fix for that issue.

eerhardt commented on src/Components/Aspire.Confluent.Kafka/MetricsService.cs:
```diff
@@ -0,0 +1,78 @@
+                }
+                catch
+                {
+                    return;
+                }
+
+                if (statistics == null || statistics.Name == null)
+                {
+                    return;
```
Do we really want to stop the MetricsService if one bad JSON document got in the channel?

Do we want to log this?

g7ed6e replied:
done. I added a warning log and replaced the return by a continue statement

eerhardt commented on src/Components/Aspire.Confluent.Kafka/MetricsService.cs:
```diff
@@ -0,0 +1,78 @@
+
+namespace Aspire.Confluent.Kafka;
+
+internal sealed class MetricsService(MetricsChannel channel, ConfluentKafkaMetrics metrics) : BackgroundService
```
Can we add some basic summary comments about what these channel, service, etc classes are/do? So someone can get a quick summary of what they represent.

g7ed6e replied:
done

eerhardt commented on src/Components/Aspire.Confluent.Kafka/AspireKafkaConsumerExtensions.cs:
```diff
@@ -0,0 +1,175 @@
+
+                // StatisticsHandler is called on the consume thread, so we need to offload the processing
+                // to avoid slowing the consumer down.
+                channel.Writer.TryWrite(json);
```
If the Channel is full, we just drop the statistic? Is that the intention?

g7ed6e replied:
Yes. The current size is 10_000. The size has been fixed to avoid a potential memory leak which may occur by creating an unbounded channel.

eerhardt reviewed (commented):
This is looking pretty good. Thanks for the contribution. I have a few comments/questions. Once they are addressed I believe we can get this merged.

mitchdenny commented on src/Aspire.Hosting/Kafka/KafkaBuilderExtensions.cs:
```diff
@@ -0,0 +1,76 @@
+    {
+        var kafka = new KafkaContainerResource(name);
+        return builder.AddResource(kafka)
+             .WithAnnotation(new EndpointAnnotation(ProtocolType.Tcp, port: port, containerPort: KafkaBrokerPort))
```
Why not use `WithEndpoint(...)`?

g7ed6e replied:
done! thanks!

mitchdenny commented on tests/Aspire.Hosting.Tests/ManifestGenerationTests.cs:
```diff
@@ -331,6 +331,29 @@ public void EnsureAllRabitMQManifestTypesHaveVersion0Suffix()
     }
 
+    [Fact]
+    public void EnsureAllKafkaManifestTypesHaveVersion0Suffix()
```
Do we need a test to makes sure the `connectionString` field is being emitted.

mitchdenny commented:
I think this is really close. A question or two from my perspective, but from an app model point of view this looks good.

g7ed6e commented on src/Components/Aspire.Confluent.Kafka/AspireKafkaConsumerExtensions.cs:
```diff
@@ -0,0 +1,175 @@
+        try
+        {
+            void OnLog(IConsumer<TKey, TValue> _, LogMessage logMessage) =>
+                logger.Log((LogLevel)logMessage.LevelAs(LogLevelType.MicrosoftExtensionsLogging), 0, logMessage.Message, null, static (value, ex) => value);
```
kafka logs does not expose an ID but a `Facility` property (string) referring to a section of librdkafka code that produced the error. Please see https://docs.confluent.io/platform/current/clients/confluent-kafka-dotnet/_site/api/Confluent.Kafka.LogMessage.html

g7ed6e replied:
updated the code set EventId to `logMessage?.Facility.GetHashCode() ?? 0`

g7ed6e commented on src/Components/Aspire.Confluent.Kafka/KafkaConsumerSettings.cs:
```diff
@@ -0,0 +1,61 @@
+            throw new InvalidOperationException("No bootstrap servers configured.");
+        }
+
+        if (Config.GroupId is null)
```
`GroupId` refers to the kafka consumer group. The broker uses this ID to track the consumed offsets by a given consumer, thus this is a required configuration parameter at consumer side.

g7ed6e commented on src/Components/Aspire.Confluent.Kafka/README.md:
````diff
@@ -0,0 +1,155 @@
+
+```csharp
+builder.AddKafkaConsumer("messaging");
+```
````
it does. good catch thanks.

g7ed6e commented on tests/Aspire.Confluent.Kafka.Tests/ProducerConfigurationTests.cs:
```diff
@@ -0,0 +1,151 @@
+
+public class ProducerConfigurationTests
+{
+    [ConditionalTheory]
```
fixed

g7ed6e commented on tests/Aspire.Hosting.Tests/Kafka/KafkaFunctionalTests.cs:
```diff
@@ -0,0 +1,36 @@
+        Assert.Equal("100", await response.Content.ReadAsStringAsync());
+
+        response = await testProgram.IntegrationServiceABuilder!.HttpGetAsync(client, "http", $"/kafka/consume/{topic}", cts.Token);
+        Assert.Equal("100", await response.Content.ReadAsStringAsync());
```
I updated tests

g7ed6e commented on src/Aspire.Hosting/Kafka/KafkaBuilderExtensions.cs:
```diff
@@ -0,0 +1,76 @@
+                int hostPort = context.PublisherName == "manifest"
+                                    ? KafkaBrokerPort
+                                    : kafka.GetPort();
+                context.EnvironmentVariables.Add("KAFKA_ADVERTISED_LISTENERS", $"PLAINTEXT://localhost:29092,PLAINTEXT_HOST://localhost:{hostPort}");
```
done

eerhardt commented on src/Components/Aspire.Confluent.Kafka/MetricsService.cs:
```diff
@@ -0,0 +1,83 @@
+        }
+    }
+
+    [LoggerMessage(LogLevel.Warning, EventId = 42, Message = "Invalid statistics json payload received: `{json}`")]
```
Why `42`?

g7ed6e replied:
which value would you suggest ?

eerhardt replied:
Looking at other usages around dotnet/aspnetcore, it looks like the convention is to start at `1`.

eerhardt commented on src/Components/Aspire.Confluent.Kafka/MetricsService.cs:
```diff
@@ -0,0 +1,83 @@
+    }
+
+    [LoggerMessage(LogLevel.Warning, EventId = 42, Message = "Invalid statistics json payload received: `{json}`")]
+    private partial void LogDeserializationWarning(string json);
```
This is kind of a bug in VS, but I'm seeing a warning that `_logger` isn't used in VS.

![image](https://github.com/dotnet/aspire/assets/8291187/33eac3e3-d916-4a3a-ab0c-fd878179e10e)

Because the _logger field is used in the generated code.

Can you change this to be

```suggestion
    private static partial void LogDeserializationWarning(ILogger logger, string json);
```

And explicitly pass the logger in? That way it is more obvious to readers how this works, and it works around the VS bug.

g7ed6e replied:
done

eerhardt commented on src/Components/Aspire.Confluent.Kafka/README.md:
````diff
@@ -0,0 +1,157 @@
+In your AppHost project, register an Apache Kafka container and consume the connection using the following methods:
+
+```csharp
+var messaging = builder.AddKafkaContainer("messaging");
````
```suggestion
var messaging = builder.AddKafka("messaging");
```

In general, we recommend the non-container resource.

g7ed6e replied:
done

eerhardt commented on tests/Aspire.Confluent.Kafka.Tests/MetricsTests.cs:
```diff
@@ -0,0 +1,345 @@
+
+public class MetricsTests
+{
+    [ConditionalTheory]
```
```suggestion
    [Theory]
```

For all the tests in this file.

g7ed6e replied:
done

eerhardt commented on tests/Aspire.Confluent.Kafka.Tests/MetricsTests.cs:
```diff
@@ -0,0 +1,345 @@
+    public record TestVariationData
+    {
+        public bool UseKeyed { get; set; }
+        public List<string>? Statistics { get; set; }
```
```suggestion
        public required List<string> Statistics { get; set; }
```

Then you don't need to `!` operator on where this is used.

g7ed6e replied:
done

eerhardt commented on tests/Aspire.Confluent.Kafka.Tests/ProducerConfigurationTests.cs:
```diff
@@ -0,0 +1,151 @@
+        Assert.Equal(CommonHelpers.TestingEndpoint, config.BootstrapServers);
+    }
+
+    [ConditionalTheory]
```
```suggestion
    [Theory]
```

For all the tests in this file.

g7ed6e replied:
done

eerhardt commented on src/Components/Aspire.Confluent.Kafka/AspireKafkaConsumerExtensions.cs:
```diff
@@ -0,0 +1,175 @@
+
+        ConsumerConfig config = new();
+        configSection.GetSection(nameof(KafkaConsumerSettings.Config)).Bind(config);
+        settings.Config = config;
```
Looking more at Kafka, there is a bit of a special pattern here in that the `ConsumerConfig` is separate from the `ConsumerBuilder`. We don't really have that pattern in the other components - typically there is just 1 object - ConfigurationOptions in Redis, IConnectionFactory in RabbitMQ.

Maybe what we have here is fine for now. It would be awkward to have 3 "configure" callbacks - configureSettings, configureConfig, configureBuilder.

g7ed6e replied:
@eerhardt The reference to the underlying library configuration object in the Aspire's "Settings" object lets customer configure a dedicated serializer / deserialize per pair of generic parameters provided to the underlying confluent XXXBuilder. Using IOptions pattern here would defeat it.
I "reproduced" the pattern I observed in the RabbitMQ implementation.

g7ed6e commented on samples/kafkaBasic/KafkaBasic.AppHost/KafkaBasic.AppHost.csproj:
```diff
@@ -0,0 +1,17 @@
+<Project Sdk="Microsoft.NET.Sdk">
```
dumped sample source code [here](https://github.com/dotnet/aspire-samples/pull/85)

g7ed6e commented on src/Aspire.Hosting/Kafka/KafkaServerResource.cs:
```diff
@@ -0,0 +1,35 @@
+namespace Aspire.Hosting.ApplicationModel;
+
+/// <summary>
+/// A resource that represents a Kafka broker connection.
```
done

g7ed6e commented:
@eerhardt I just squashed end rebased until the Orleans support adding.

eerhardt committed:
Address PR feedback

- Make Kafka Settings Config object non-nullable and not settable from user code.
- Use EventId 1 in Logging
- Log response body in Functional tests

eerhardt reviewed (approved):
This is great! Thanks, @g7ed6e, for this contribution and all your great work.

I pushed some review feedback changes. Let me know if you have any issues with the changes.

Unless anyone else has any objections, I will merge this once CI passes.

eerhardt committed:
Adjust the initialization of the Settings Config properties because the type can be created publicly.

davidfowl commented:
Thanks so much @g7ed6e!
